File: Fortran/Persistent/Failures/warning.f90
PROGRAM dch

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100
  INTEGER, PARAMETER :: n_elements = 10

  INTEGER :: rank
  INTEGER :: nproc
  INTEGER :: left, right
  INTEGER :: ierr
  INTEGER, DIMENSION(2) :: requests
  INTEGER, DIMENSION(MPI_STATUS_SIZE, 2) :: statuses
  INTEGER :: repcount
  REAL :: time1, time2, time3, dtime1, dtime2
  INTEGER, DIMENSION(:), ALLOCATABLE :: send, recv

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !Set up periodic domain
  left = rank - 1
  IF (left < 0) left = nproc - 1
  right = rank + 1
  IF (right > nproc - 1) right = 0

  time1 = MPI_Wtime()
  !If this isn't allocated here then MPI_Send_init detects a runtime error
  !still doesn't work
  ALLOCATE(send(n_elements), recv(n_elements))
  CALL MPI_Send_init(send, n_elements, MPI_INTEGER, right, tag, MPI_COMM_WORLD,&
    requests(1), ierr)
  CALL MPI_Recv_init(recv, n_elements, MPI_INTEGER, left , tag, MPI_COMM_WORLD,&
    requests(2), ierr)
  DEALLOCATE(send, recv)
  time2 = MPI_Wtime()
  DO repcount = 1, 1000000
    !Each cycle, reallocate your send and receive buffers
    !In general this won't be where it was last time in memory
    !It *might* be in which case this will work, but only by coincidence.
    !This is intended to mimic the idea of using temporary arrays
    ALLOCATE(send(n_elements), recv(n_elements))
    send = rank
    CALL MPI_Startall(2, requests, ierr)
    CALL MPI_Waitall(2, requests, statuses, ierr)
    !And then deallocate them
    DEALLOCATE(send, recv)
  END DO
  time3 = MPI_Wtime()

  dtime1 = time3-time1
  CALL MPI_Reduce(dtime1, MPI_IN_PLACE, 1, MPI_REAL, MPI_MAX, 0, &
      MPI_COMM_WORLD, ierr)
  dtime2 = time2 - time1
  CALL MPI_Reduce(dtime2, MPI_IN_PLACE, 1, MPI_REAL, MPI_MAX, 0, &
      MPI_COMM_WORLD, ierr)

  PRINT *,"Rank ", rank, " got message from rank ", left, " of ", MAXVAL(recv),&
      "should be ", left
  CALL MPI_Barrier(MPI_COMM_WORLD, ierr)
  IF (rank == 0) PRINT *, "It took ", dtime1, " seconds to complete send/receives"
  IF (rank == 0) PRINT *, "It took ", dtime2, " seconds to complete setup"

  CALL MPI_Finalize(ierr)

END PROGRAM dch

File: C/Persistent/Failures/warning.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define TAG 100
#define NELEMENTS 10

int main(int argc, char ** argv)
{

  int rank, maxval, nproc, left, right, repcount, cycle;
  double time1, time2, time3, dtime1, dtime2, dtime1_r, dtime2_r;
  MPI_Request requests[2];
  MPI_Status statuses[2];
  //Dynamic arrays this time
  int *send, *recv;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  time1 = MPI_Wtime();
  //In Fortran, MPI can detect if an array is allocated or not so allocate here
  //Only here to keep C and Fortran codes similar
  send = (int*)malloc(sizeof(int)*NELEMENTS);
  recv = (int*)malloc(sizeof(int)*NELEMENTS);

  for(cycle =0 ;cycle<NELEMENTS;++cycle){
    send[cycle] = rank;
  }

  MPI_Send_init(send, NELEMENTS, MPI_INT, right, TAG, MPI_COMM_WORLD,
      requests);
  MPI_Recv_init(recv, NELEMENTS, MPI_INT, left , TAG, MPI_COMM_WORLD,
      requests+1);
  time2 = MPI_Wtime();
  free(send);
  free(recv);
  for (repcount = 0; repcount <100000; ++repcount) {
    //Each cycle, reallocate your send and receive buffers
    //In general this won't be where it was last time in memory
    //It *might* be in which case this will work, but only by coincidence.
    //This is intended to mimic the idea of using temporary arrays
    send = (int*)malloc(sizeof(int)*NELEMENTS);
    recv = (int*)malloc(sizeof(int)*NELEMENTS);
    MPI_Startall(2, requests);
    MPI_Waitall(2, requests, statuses);
    //Then deallocate
    free(send);
    free(recv);
  }
  time3 = MPI_Wtime();

  dtime1 = time3 - time1;
  MPI_Reduce(&dtime1, &dtime1_r, 1, MPI_DOUBLE, MPI_MAX, 0,
      MPI_COMM_WORLD);

  dtime2 = time3 - time2;
  MPI_Reduce(&dtime2, &dtime2_r, 1, MPI_DOUBLE, MPI_MAX, 0,
      MPI_COMM_WORLD);

  maxval = 0;
  for (cycle = 0; cycle<NELEMENTS; ++cycle){
    if (recv[cycle] > maxval) maxval = recv[cycle];
  }

  printf("Rank %3d got value from rank %3d of %3d. Should be %3d\n",
      rank, left, maxval, left);
  MPI_Barrier(MPI_COMM_WORLD);
  if (rank == 0) {
    printf("It took %f seconds to complete send/receives\n", dtime1_r);
    printf("It took %f seconds to to complete setup\n", dtime2_r);
  }

  MPI_Finalize();

}
